### ISSUE TITLE
Worker processes end up spending all their time in GC due to metric tag allocations

### ISSUE THREAD

**Issue by smcgivern** [temporalio/temporal#4716]
We run Temporal 1.20.3 on Kubernetes. It has very light load. We have noticed that over time, the worker process consumes more and more CPU time and memory, eventually getting 'stuck' and requiring a restart. Here's a chart from GCP showing CPU and memory over a one-week period in July (blue line is usage, green is requests, red is limits). You can see that we restarted this manually at the end.

<img width="796" alt="image" src="https://github.com/temporalio/temporal/assets/1120328/4dd3149f-0d36-477f-97e6-d4af2a1f8944">

Additionally, a full metrics scrape against one of these pods (using `kubectl port-forward`) takes 7 seconds, even in the idle state:

```shell
$ time curl -I localhost:9090/metrics
HTTP/1.1 200 OK
Content-Type: text/plain; version=0.0.4; charset=utf-8
Date: Wed, 02 Aug 2023 10:30:32 GMT


real	0m7.580s
user	0m0.006s
sys	0m0.011s
```

## Profiles

These profiles were taken when CPU usage and memory usage are very high. 

* [pprof.temporal-server.alloc_objects.alloc_space.inuse_objects.inuse_space.001.pb.gz](https://github.com/temporalio/temporal/files/12239105/pprof.temporal-server.alloc_objects.alloc_space.inuse_objects.inuse_space.001.pb.gz)
* [pprof.temporal-server.samples.cpu.001.pb.gz](https://github.com/temporalio/temporal/files/12239107/pprof.temporal-server.samples.cpu.001.pb.gz)

The memory profile shows - make sure to change to `alloc_space` sampling to include all allocations, not just live memory - that the majority of the allocations were in `metrics.tagsToAttributes`:

<img width="1348" alt="image" src="https://github.com/temporalio/temporal/assets/1120328/9635ad82-0445-4a0a-ab40-5b8d1bc12651">

That also shows in the CPU profile, but not as dramatically. We can see a lot of GC time:

<img width="1348" alt="image" src="https://github.com/temporalio/temporal/assets/1120328/be44b603-cc98-4a2b-9712-e15aa306af93">

## Config

This is our metrics config:

```yaml
metrics:
  tags:
    type: "worker"
  prometheus:
    framework: "opentelemetry"
    timerType: "histogram"
    listenAddress: "0.0.0.0:9090"
```

And we give it these resources:

```yaml
resources:
  limits:
    cpu: 800m
    memory: 2000Mi
    ephemeral-storage: "1Gi"
  requests:
    cpu: 500m
    memory: 500Mi
    ephemeral-storage: "500Mi"
```

## Mitigation

We changed our scrape interval from 15 seconds to 60 seconds and it now takes much longer to get in this state, as expected. It still ends up there eventually, though.

## Side note

It seems like the dynamic metric tags come from two places:

1. Config - this doesn't change during runtime.
2. Call sites - this doesn't change after compilation.

So recalculating this every time does seem inefficient, but I'm not sure why it's not a problem for other people and it is for us.

## Related issues and PRs

* https://github.com/temporalio/temporal/pull/3578 looks like it improved this but it was closed. I'm not clear on why - @wxing1292, would you mind elaborating?
* https://github.com/temporalio/temporal/issues/3015 is about `tagsToMap` which is for Tally metrics and seems to do a similar thing

**Comment by smcgivern** [comment]
I was able to reproduce some very slow responses with these changes to https://github.com/tsurdilo/my-temporal-dockercompose:

```diff
diff --git a/compose-postgres.yml b/compose-postgres.yml
index 400491a..5b65488 100644
--- a/compose-postgres.yml
+++ b/compose-postgres.yml
@@ -7,7 +7,7 @@ services:
       POSTGRES_USER: temporal
     image: postgres:13
     ports:
-      - 5432:5432
+      - 5444:5432
   postgres-exporter:
     container_name: postgres-exporter
     image: wrouesnel/postgres_exporter:v0.8.0
diff --git a/compose-services.yml b/compose-services.yml
index 44cdbc9..6539585 100644
--- a/compose-services.yml
+++ b/compose-services.yml
@@ -219,6 +219,7 @@ services:
       - DYNAMIC_CONFIG_FILE_PATH=config/dynamicconfig/development.yaml
       - SERVICES=worker
       - PROMETHEUS_ENDPOINT=0.0.0.0:8003
+      - PPROF_PORT=8989
       # set to loadbalancing
       - USE_INTERNAL_FRONTEND=true
     image: temporalio/server:${TEMPORAL_SERVER_IMG}
@@ -227,6 +228,7 @@ services:
         target: 7232
       - published: 8003
         target: 8003
+      - 8989:8989
     restart: on-failure
     volumes:
       - ./dynamicconfig:/etc/temporal/config/dynamicconfig
diff --git a/deployment/prometheus/config.yml b/deployment/prometheus/config.yml
index bcbd173..47acae1 100644
--- a/deployment/prometheus/config.yml
+++ b/deployment/prometheus/config.yml
@@ -4,6 +4,7 @@ scrape_configs:
   - job_name: 'temporalmetrics'
     metrics_path: /metrics
     scheme: http
+    scrape_interval: 1s
     static_configs:
       # Server metrics target
       - targets:
diff --git a/template/my_config_template.yaml b/template/my_config_template.yaml
index 399bb96..18c6b72 100644
--- a/template/my_config_template.yaml
+++ b/template/my_config_template.yaml
@@ -279,6 +279,8 @@ global:
       prefix: "temporal"
   {{- else if .Env.PROMETHEUS_ENDPOINT }}
   metrics:
+    tags:
+      some_label: value
     prometheus:
       timerType: {{ default .Env.PROMETHEUS_TIMER_TYPE "histogram" }}
       listenAddress: "{{ .Env.PROMETHEUS_ENDPOINT }}"
```

For instance, not long after starting:

```shell
$ time curl -I localhost:8003/metrics
HTTP/1.1 200 OK
Content-Type: text/plain; version=0.0.4; charset=utf-8
Date: Wed, 02 Aug 2023 10:41:18 GMT


real	0m0.022s
user	0m0.005s
sys	0m0.009s
```

And then later:

```shell
$ time curl -I localhost:8003/metrics
HTTP/1.1 200 OK
Content-Type: text/plain; version=0.0.4; charset=utf-8
Date: Wed, 02 Aug 2023 11:22:35 GMT


real	0m58.154s
user	0m0.003s
sys	0m0.010s

**Comment by smcgivern** [comment]
@dnr looks like https://github.com/temporalio/temporal/pull/4722 was merged :tada:

Does that close this issue?

**Comment by dnr** [comment]
> Does that close this issue?

I hope so! We can keep it open until it's confirmed though. That PR will be released with upcoming server version 1.22

**Comment by smcgivern** [comment]
This is what memory usage looked like for the past two weeks in our deployment. I think it's very clear where we switched to 1.22 - thanks for your help here, @dnr!

<img width="1478" alt="image" src="https://github.com/temporalio/temporal/assets/1120328/4a5807dc-2951-485b-beae-336aa8d4d426">
